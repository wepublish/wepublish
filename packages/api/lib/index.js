"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getUserForCredentials = void 0;
__exportStar(require("./graphql/session"), exports);
__exportStar(require("./graphql/navigation"), exports);
__exportStar(require("./graphql/author"), exports);
__exportStar(require("./graphql/image"), exports);
__exportStar(require("./graphql/blocks"), exports);
__exportStar(require("./graphql/article"), exports);
__exportStar(require("./graphql/comment/comment"), exports);
__exportStar(require("./graphql/page"), exports);
__exportStar(require("./graphql/peer"), exports);
__exportStar(require("./graphql/token"), exports);
__exportStar(require("./graphql/richText"), exports);
__exportStar(require("./graphql/slug"), exports);
// export * from './graphql/mutation.private'
// export * from './graphql/query.private'
__exportStar(require("./graphql/schema"), exports);
__exportStar(require("./graphql/permissions"), exports);
__exportStar(require("./db/user"), exports);
__exportStar(require("./db/userRole"), exports);
__exportStar(require("./db/session"), exports);
__exportStar(require("./db/navigation"), exports);
__exportStar(require("./db/author"), exports);
__exportStar(require("./db/image"), exports);
__exportStar(require("./db/block"), exports);
__exportStar(require("./db/article"), exports);
__exportStar(require("./db/comment"), exports);
__exportStar(require("./db/page"), exports);
__exportStar(require("./db/common"), exports);
__exportStar(require("./db/peer"), exports);
__exportStar(require("./db/memberPlan"), exports);
__exportStar(require("./db/invoice"), exports);
__exportStar(require("./db/payment"), exports);
__exportStar(require("./db/mailLog"), exports);
__exportStar(require("./db/subscription"), exports);
__exportStar(require("./db/setting"), exports);
__exportStar(require("./media/mediaAdapter"), exports);
__exportStar(require("./media/karmaMediaAdapter"), exports);
__exportStar(require("./urlAdapter"), exports);
__exportStar(require("./utility"), exports);
__exportStar(require("./error"), exports);
__exportStar(require("./jobs"), exports);
__exportStar(require("./context"), exports);
__exportStar(require("./server"), exports);
__exportStar(require("./runServer"), exports);
__exportStar(require("./payments/paymentProvider"), exports);
__exportStar(require("./payments/stripeCheckoutPaymentProvider"), exports);
__exportStar(require("./payments/stripePaymentProvider"), exports);
__exportStar(require("./payments/payrexxPaymentProvider"), exports);
__exportStar(require("./mails/mailProvider"), exports);
__exportStar(require("./mails/MailgunMailProvider"), exports);
__exportStar(require("./mails/MailchimpMailProvider"), exports);
__exportStar(require("./mails/mailContext"), exports);
__exportStar(require("./challenges/challengeProvider"), exports);
__exportStar(require("./challenges/algebraicCaptchaChallenge"), exports);
__exportStar(require("./payments/payrexxSubscriptionPaymentProvider"), exports);
var user_queries_1 = require("./graphql/user/user.queries");
Object.defineProperty(exports, "getUserForCredentials", { enumerable: true, get: function () { return user_queries_1.getUserForCredentials; } });
//# sourceMappingURL=index.js.map