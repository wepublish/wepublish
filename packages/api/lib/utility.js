"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkSettingRestrictions = exports.isBoolean = exports.isString = exports.isArray = exports.isObject = exports.countRichtextChars = exports.capitalizeFirstLetter = exports.delegateToPeerSchema = exports.mapEnumsBack = exports.createProxyingIsTypeOf = exports.createProxyingResolver = exports.isSourceProxied = exports.markResultAsProxied = exports.IsProxiedSymbol = exports.base64Decode = exports.base64Encode = exports.slugify = exports.mapSubscriptionsAsCsv = exports.USER_PROPERTY_LAST_LOGIN_LINK_SEND = exports.ONE_MONTH_IN_MILLISECONDS = exports.FIFTEEN_MINUTES_IN_MILLISECONDS = exports.ONE_DAY_IN_MILLISECONDS = exports.ONE_HOUR_IN_MILLISECONDS = exports.MAX_PAYLOAD_SIZE = exports.MAX_COMMENT_LENGTH = void 0;
const formatISO_1 = __importDefault(require("date-fns/formatISO"));
const graphql_tools_1 = require("graphql-tools");
const block_1 = require("./db/block");
const error_1 = require("./error");
exports.MAX_COMMENT_LENGTH = 1000;
exports.MAX_PAYLOAD_SIZE = '1MB';
exports.ONE_HOUR_IN_MILLISECONDS = 60 * 60 * 1000;
exports.ONE_DAY_IN_MILLISECONDS = 24 * exports.ONE_HOUR_IN_MILLISECONDS;
exports.FIFTEEN_MINUTES_IN_MILLISECONDS = 900000;
exports.ONE_MONTH_IN_MILLISECONDS = 31 * exports.ONE_DAY_IN_MILLISECONDS;
exports.USER_PROPERTY_LAST_LOGIN_LINK_SEND = '_wepLastLoginLinkSentTimestamp';
function mapSubscriptionsAsCsv(subscriptions) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p;
    let csvStr = [
        'id',
        'firstName',
        'name',
        'preferredName',
        'email',
        'active',
        'createdAt',
        'modifiedAt',
        'company',
        'streetAddress',
        'streetAddress2',
        'zipCode',
        'city',
        'country',
        'memberPlan',
        'memberPlanID',
        'paymentPeriodicity',
        'monthlyAmount',
        'autoRenew',
        'startsAt',
        'paidUntil',
        'paymentMethod',
        'paymentMethodID',
        'deactivationDate',
        'deactivationReason'
    ].join(',') + '\n';
    for (const subscription of subscriptions) {
        const user = subscription === null || subscription === void 0 ? void 0 : subscription.user;
        const memberPlan = subscription === null || subscription === void 0 ? void 0 : subscription.memberPlan;
        const paymentMethod = subscription === null || subscription === void 0 ? void 0 : subscription.paymentMethod;
        // if (!user) continue
        csvStr +=
            [
                user === null || user === void 0 ? void 0 : user.id,
                `${sanitizeCsvContent(user === null || user === void 0 ? void 0 : user.firstName)}`,
                `${sanitizeCsvContent(user === null || user === void 0 ? void 0 : user.name)}`,
                `${sanitizeCsvContent(user === null || user === void 0 ? void 0 : user.preferredName)}`,
                `${(_a = user === null || user === void 0 ? void 0 : user.email) !== null && _a !== void 0 ? _a : ''}`,
                user === null || user === void 0 ? void 0 : user.active,
                (user === null || user === void 0 ? void 0 : user.createdAt) ? (0, formatISO_1.default)(user.createdAt, { representation: 'date' }) : '',
                (user === null || user === void 0 ? void 0 : user.modifiedAt) ? (0, formatISO_1.default)(user.modifiedAt, { representation: 'date' }) : '',
                `${sanitizeCsvContent((_b = user === null || user === void 0 ? void 0 : user.address) === null || _b === void 0 ? void 0 : _b.company)}`,
                `${sanitizeCsvContent((_c = user === null || user === void 0 ? void 0 : user.address) === null || _c === void 0 ? void 0 : _c.streetAddress)}`,
                `${sanitizeCsvContent((_d = user === null || user === void 0 ? void 0 : user.address) === null || _d === void 0 ? void 0 : _d.streetAddress2)}`,
                `${sanitizeCsvContent((_e = user === null || user === void 0 ? void 0 : user.address) === null || _e === void 0 ? void 0 : _e.zipCode)}`,
                `${sanitizeCsvContent((_f = user === null || user === void 0 ? void 0 : user.address) === null || _f === void 0 ? void 0 : _f.city)}`,
                `${sanitizeCsvContent((_g = user === null || user === void 0 ? void 0 : user.address) === null || _g === void 0 ? void 0 : _g.country)}`,
                sanitizeCsvContent(memberPlan === null || memberPlan === void 0 ? void 0 : memberPlan.name),
                (_h = subscription === null || subscription === void 0 ? void 0 : subscription.memberPlanID) !== null && _h !== void 0 ? _h : '',
                (_j = subscription === null || subscription === void 0 ? void 0 : subscription.paymentPeriodicity) !== null && _j !== void 0 ? _j : '',
                (_k = subscription === null || subscription === void 0 ? void 0 : subscription.monthlyAmount) !== null && _k !== void 0 ? _k : '',
                (_l = subscription === null || subscription === void 0 ? void 0 : subscription.autoRenew) !== null && _l !== void 0 ? _l : '',
                (subscription === null || subscription === void 0 ? void 0 : subscription.startsAt) ? (0, formatISO_1.default)(subscription.startsAt, { representation: 'date' }) : '',
                (subscription === null || subscription === void 0 ? void 0 : subscription.paidUntil)
                    ? (0, formatISO_1.default)(subscription.paidUntil, { representation: 'date' })
                    : 'no pay',
                sanitizeCsvContent(paymentMethod === null || paymentMethod === void 0 ? void 0 : paymentMethod.name),
                (_m = subscription === null || subscription === void 0 ? void 0 : subscription.paymentMethodID) !== null && _m !== void 0 ? _m : '',
                (subscription === null || subscription === void 0 ? void 0 : subscription.deactivation)
                    ? (0, formatISO_1.default)(subscription.deactivation.date, { representation: 'date' })
                    : '',
                (_p = (_o = subscription === null || subscription === void 0 ? void 0 : subscription.deactivation) === null || _o === void 0 ? void 0 : _o.reason) !== null && _p !== void 0 ? _p : ''
            ].join(',') + '\r\n';
    }
    return csvStr;
}
exports.mapSubscriptionsAsCsv = mapSubscriptionsAsCsv;
/**
 * according to rfc 4180
 * https://www.ietf.org/rfc/rfc4180.txt
 * @param input
 */
function sanitizeCsvContent(input) {
    // according rfc 4180 2.7.
    const escapeDoubleQuotes = (input || '').toString().replace(/[#"]/g, '""');
    // according rfc 4180 2.5. / 2.6.
    return `"${escapeDoubleQuotes}"`;
}
// https://gist.github.com/mathewbyrne/1280286#gistcomment-2588056
function slugify(str) {
    return str
        .toLowerCase()
        .trim()
        .replace(/[ÀÁÂÃÄÅÆĀĂĄẠẢẤẦẨẪẬẮẰẲẴẶ]/gi, 'a')
        .replace(/[ÇĆĈČ]/gi, 'c')
        .replace(/[ÐĎĐÞ]/gi, 'd')
        .replace(/[ÈÉÊËĒĔĖĘĚẸẺẼẾỀỂỄỆ]/gi, 'e')
        .replace(/[ĜĞĢǴ]/gi, 'g')
        .replace(/[ĤḦ]/gi, 'h')
        .replace(/[ÌÍÎÏĨĪĮİỈỊ]/gi, 'i')
        .replace(/[Ĵ]/gi, 'j')
        .replace(/[Ĳ]/gi, 'ij')
        .replace(/[Ķ]/gi, 'k')
        .replace(/[ĹĻĽŁ]/gi, 'l')
        .replace(/[Ḿ]/gi, 'm')
        .replace(/[ÑŃŅŇ]/gi, 'n')
        .replace(/[ÒÓÔÕÖØŌŎŐỌỎỐỒỔỖỘỚỜỞỠỢǪǬƠ]/gi, 'o')
        .replace(/[Œ]/gi, 'oe')
        .replace(/[ṕ]/gi, 'p')
        .replace(/[ŔŖŘ]/gi, 'r')
        .replace(/[ßŚŜŞŠ]/gi, 's')
        .replace(/[ŢŤ]/gi, 't')
        .replace(/[ÙÚÛÜŨŪŬŮŰŲỤỦỨỪỬỮỰƯ]/gi, 'u')
        .replace(/[ẂŴẀẄ]/gi, 'w')
        .replace(/[ẍ]/gi, 'x')
        .replace(/[ÝŶŸỲỴỶỸ]/gi, 'y')
        .replace(/[ŹŻŽ]/gi, 'z')
        .replace(/[·/_,:;\\']/gi, '-')
        .replace(/\s+/g, '-')
        .replace(/[^\w\-]+/g, '')
        .replace(/--+/g, '-')
        .replace(/^-+/, '')
        .replace(/-+$/, '');
}
exports.slugify = slugify;
function base64Encode(str) {
    return Buffer.from(str).toString('base64');
}
exports.base64Encode = base64Encode;
function base64Decode(str) {
    return Buffer.from(str, 'base64').toString();
}
exports.base64Decode = base64Decode;
exports.IsProxiedSymbol = Symbol('isProxied');
function markResultAsProxied(result) {
    if (!result)
        return null;
    for (const key in result) {
        const value = result[key];
        if (typeof value === 'object' && value != null) {
            markResultAsProxied(value);
        }
    }
    return Object.assign(result, { [exports.IsProxiedSymbol]: true });
}
exports.markResultAsProxied = markResultAsProxied;
function isSourceProxied(source) {
    return source[exports.IsProxiedSymbol] == true;
}
exports.isSourceProxied = isSourceProxied;
function createProxyingResolver(resolver) {
    return (source, args, context, info) => {
        if (isSourceProxied(source)) {
            return source[info.path.key];
        }
        return resolver(source, args, context, info);
    };
}
exports.createProxyingResolver = createProxyingResolver;
function createProxyingIsTypeOf(isTypeOf) {
    return function (source, context, info) {
        return isSourceProxied(source)
            ? source.__typename === this.name
            : isTypeOf(source, context, info);
    };
}
exports.createProxyingIsTypeOf = createProxyingIsTypeOf;
function mapEnumsBack(result) {
    if (!result)
        return null;
    for (const key in result) {
        const value = result[key];
        if (typeof value === 'object' && value !== null) {
            mapEnumsBack(value);
        }
    }
    if (result.__typename === 'ArticleTeaser' ||
        result.__typename === 'PeerArticleTeaser' ||
        result.__typename === 'PageTeaser' ||
        result.__typename === 'CustomTeaser') {
        switch (result.style) {
            case 'DEFAULT':
                return Object.assign(result, { style: block_1.TeaserStyle.Default });
            case 'LIGHT':
                return Object.assign(result, { style: block_1.TeaserStyle.Light });
            case 'TEXT':
                return Object.assign(result, { style: block_1.TeaserStyle.Text });
        }
    }
    return result;
}
exports.mapEnumsBack = mapEnumsBack;
class ResetGraphQLEnums {
    transformResult(result) {
        // FIXME: WPC-415 created
        return mapEnumsBack(result);
    }
}
async function delegateToPeerSchema(peerID, fetchAdminEndpoint, context, opts) {
    var _a;
    const schema = fetchAdminEndpoint
        ? await context.loaders.peerAdminSchema.load(peerID)
        : await context.loaders.peerSchema.load(peerID);
    if (!schema)
        return null;
    return markResultAsProxied(await (0, graphql_tools_1.delegateToSchema)(Object.assign(Object.assign({}, opts), { schema, transforms: [new ResetGraphQLEnums(), ...((_a = opts.transforms) !== null && _a !== void 0 ? _a : [])] })));
}
exports.delegateToPeerSchema = delegateToPeerSchema;
function capitalizeFirstLetter(str) {
    return str.charAt(0).toUpperCase() + str.slice(1);
}
exports.capitalizeFirstLetter = capitalizeFirstLetter;
function countRichtextChars(blocksCharLength, nodes) {
    return nodes.reduce((charLength, node) => {
        if (!node.text && !node.children)
            return charLength;
        if (node.text) {
            return charLength + node.text.length;
        }
        return countRichtextChars(charLength, node.children);
    }, blocksCharLength);
}
exports.countRichtextChars = countRichtextChars;
function isObject(unknown) {
    return typeof unknown === 'object' && unknown != null && !Array.isArray(unknown);
}
exports.isObject = isObject;
function isArray(unknown) {
    return Array.isArray(unknown);
}
exports.isArray = isArray;
function isString(unknown) {
    return typeof unknown === 'string';
}
exports.isString = isString;
function isBoolean(unknown) {
    return typeof unknown === 'boolean';
}
exports.isBoolean = isBoolean;
function checkSettingRestrictions(val, currentVal, restriction) {
    var _a, _b;
    if (!restriction) {
        return;
    }
    if (typeof val !== typeof currentVal) {
        throw new error_1.InvalidSettingValueError();
    }
    if (((_a = restriction.allowedValues) === null || _a === void 0 ? void 0 : _a.boolChoice) && typeof val !== 'boolean') {
        throw new error_1.InvalidSettingValueError();
    }
    if (typeof val === 'number') {
        if (restriction.maxValue && val > restriction.maxValue) {
            throw new error_1.InvalidSettingValueError();
        }
        if (restriction.minValue && val < restriction.minValue) {
            throw new error_1.InvalidSettingValueError();
        }
    }
    if (typeof val === 'string') {
        if (restriction.inputLength && val.length > restriction.inputLength) {
            throw new error_1.InvalidSettingValueError();
        }
        if (((_b = restriction.allowedValues) === null || _b === void 0 ? void 0 : _b.stringChoice) &&
            !restriction.allowedValues.stringChoice.includes(val)) {
            throw new error_1.InvalidSettingValueError();
        }
    }
}
exports.checkSettingRestrictions = checkSettingRestrictions;
//# sourceMappingURL=utility.js.map