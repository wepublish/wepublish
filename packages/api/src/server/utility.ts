import {sign as signJWT, verify as verifyJWT} from 'jsonwebtoken'

export enum JWTAudience {
  Refresh = 'refresh',
  Access = 'access',
  Peer = 'peer'
}

export interface RefreshToken {
  email: string
  id: string
}

export function signRefreshToken(
  id: string,
  email: string,
  tokenSecret: string,
  expiresIn: number
): string {
  return signJWT({}, tokenSecret, {
    expiresIn,
    subject: `user:${email}`,
    audience: JWTAudience.Refresh,
    jwtid: id
  })
}

export function signAccessToken(email: string, tokenSecret: string, expiresIn: number): string {
  return signJWT({}, tokenSecret, {
    expiresIn,
    subject: `user:${email}`,
    audience: JWTAudience.Access
  })
}

export function verifyRefreshToken(token: string): RefreshToken {
  const tokenData = verifyJWT(token, 'secret', {
    audience: JWTAudience.Refresh
  }) as {sub: string; jti: string}

  return {
    email: tokenData.sub.split('user:')[1],
    id: tokenData.jti
  }
}
