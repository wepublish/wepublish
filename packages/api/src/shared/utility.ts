import generateNanoID from 'nanoid/async/generate'
import generateNanoIDSync from 'nanoid/generate'

export const idAlphabet = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ'
export const tokenAlphabet = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ-_'

export function generateID() {
  return generateNanoID(idAlphabet, 10)
}

export function generateToken() {
  return generateNanoID(tokenAlphabet, 32)
}

export function generateIDSync() {
  return generateNanoIDSync(idAlphabet, 10)
}

export function generateTokenSync() {
  return generateNanoIDSync(tokenAlphabet, 32)
}
