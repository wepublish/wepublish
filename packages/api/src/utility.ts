import {GraphQLFieldResolver, GraphQLIsTypeOfFn, GraphQLObjectType} from 'graphql'

import {delegateToSchema, IDelegateToSchemaOptions} from 'graphql-tools'

import {Context} from './context'

// https://gist.github.com/mathewbyrne/1280286#gistcomment-2588056
export function slugify(str: string) {
  return str
    .toLowerCase()
    .trim()
    .replace(/[ÀÁÂÃÄÅÆĀĂĄẠẢẤẦẨẪẬẮẰẲẴẶ]/gi, 'a')
    .replace(/[ÇĆĈČ]/gi, 'c')
    .replace(/[ÐĎĐÞ]/gi, 'd')
    .replace(/[ÈÉÊËĒĔĖĘĚẸẺẼẾỀỂỄỆ]/gi, 'e')
    .replace(/[ĜĞĢǴ]/gi, 'g')
    .replace(/[ĤḦ]/gi, 'h')
    .replace(/[ÌÍÎÏĨĪĮİỈỊ]/gi, 'i')
    .replace(/[Ĵ]/gi, 'j')
    .replace(/[Ĳ]/gi, 'ij')
    .replace(/[Ķ]/gi, 'k')
    .replace(/[ĹĻĽŁ]/gi, 'l')
    .replace(/[Ḿ]/gi, 'm')
    .replace(/[ÑŃŅŇ]/gi, 'n')
    .replace(/[ÒÓÔÕÖØŌŎŐỌỎỐỒỔỖỘỚỜỞỠỢǪǬƠ]/gi, 'o')
    .replace(/[Œ]/gi, 'oe')
    .replace(/[ṕ]/gi, 'p')
    .replace(/[ŔŖŘ]/gi, 'r')
    .replace(/[ßŚŜŞŠ]/gi, 's')
    .replace(/[ŢŤ]/gi, 't')
    .replace(/[ÙÚÛÜŨŪŬŮŰŲỤỦỨỪỬỮỰƯ]/gi, 'u')
    .replace(/[ẂŴẀẄ]/gi, 'w')
    .replace(/[ẍ]/gi, 'x')
    .replace(/[ÝŶŸỲỴỶỸ]/gi, 'y')
    .replace(/[ŹŻŽ]/gi, 'z')
    .replace(/[·/_,:;\\']/gi, '-')
    .replace(/\s+/g, '-')
    .replace(/[^\w\-]+/g, '')
    .replace(/--+/g, '-')
    .replace(/^-+/, '')
    .replace(/-+$/, '')
}

export function base64Encode(str: string): string {
  return Buffer.from(str).toString('base64')
}

export function base64Decode(str: string): string {
  return Buffer.from(str, 'base64').toString()
}

export const IsProxiedSymbol = Symbol('isProxied')

export function markResultAsProxied(result: any) {
  if (!result) return null

  for (const key in result) {
    const value = result[key]

    if (typeof value === 'object' && value != null) {
      markResultAsProxied(value)
    }
  }

  return Object.assign(result, {[IsProxiedSymbol]: true})
}

export function isSourceProxied<T>(source: T): source is T & {__typename?: string} {
  return (source as any)[IsProxiedSymbol] == true
}

export function createProxyingResolver<TSource, TContext, TArgs = {[argName: string]: any}>(
  resolver: GraphQLFieldResolver<TSource, TContext, TArgs>
): GraphQLFieldResolver<TSource, TContext, TArgs> {
  return (source, args, context, info) => {
    if (isSourceProxied(source)) {
      return (source as any)[info.path.key]
    }

    return resolver(source, args, context, info)
  }
}

export function createProxyingIsTypeOf<TSource, TContext>(
  isTypeOf: GraphQLIsTypeOfFn<TSource, TContext>
): GraphQLIsTypeOfFn<TSource, TContext> {
  return function (this: GraphQLObjectType, source, context, info) {
    return isSourceProxied(source)
      ? source.__typename === this.name
      : isTypeOf(source, context, info)
  }
}

export async function delegateToPeerSchema(
  peerID: string,
  fetchAdminEndpoint: boolean,
  context: Context,
  opts: Omit<IDelegateToSchemaOptions, 'schema'>
) {
  const schema = fetchAdminEndpoint
    ? await context.loaders.peerAdminSchema.load(peerID)
    : await context.loaders.peerSchema.load(peerID)

  if (!schema) return null

  return markResultAsProxied(
    await delegateToSchema({
      ...opts,
      schema: schema,
      transforms: [...(opts.transforms ?? [])]
    })
  )
}

export function capitalizeFirstLetter(str: string): string {
  return str.charAt(0).toUpperCase() + str.slice(1)
}

export const ONE_HOUR_IN_MILLISECONDS = 60 * 60 * 1000
export const ONE_DAY_IN_MILLISECONDS = 24 * ONE_HOUR_IN_MILLISECONDS
