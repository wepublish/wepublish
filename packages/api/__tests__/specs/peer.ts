import {MongoDBAdapter} from '@wepublish/api-db-mongodb'
import {ApolloServerTestClient} from 'apollo-server-testing'
import {createGraphQLTestClientWithMongoDB} from '../utility'
import {
  CreatePeerInput,
  CreatePeer,
  PeerList,
  Peer,
  PeerProfile,
  UpdatePeer,
  DeletePeer,
  UpdatePeerProfile
} from '../api/private'
import {FetchMock} from 'jest-fetch-mock'
import fetch from 'node-fetch'
import fakePeerAdminSchema from '../fakePeerAdminSchema.json'
import {PeerProfileInput} from '../../lib'

// Mocking Fetch calls from the "createFetcher" method in context
;((fetch as unknown) as FetchMock).mockResponse(JSON.stringify(fakePeerAdminSchema))

let testClientPublic: ApolloServerTestClient
let testClientPrivate: ApolloServerTestClient
let dbAdapter: MongoDBAdapter

beforeAll(async () => {
  try {
    const setupClient = await createGraphQLTestClientWithMongoDB()
    testClientPublic = setupClient.testClientPublic
    testClientPrivate = setupClient.testClientPrivate
    dbAdapter = setupClient.dbAdapter

    console.log('public', testClientPublic)
  } catch (error) {
    console.log('Error', error)

    throw new Error('Error during test setup')
  }
})

describe('Peers', () => {
  describe('can be created/edited/deleted:', () => {
    const ids: string[] = []
    beforeEach(async () => {
      const {mutate} = testClientPrivate
      const input: CreatePeerInput = {
        name: `Peer Test ${ids.length}`,
        slug: `test-peer${ids.length}`,
        hostURL: 'https://host-url.ch/',
        token: `token${ids.length}`
      }
      const res = await mutate({
        mutation: CreatePeer,
        variables: {
          input: input
        }
      })
      ids.unshift(res.data?.createPeer?.id)
    })

    test('can be created', async () => {
      const {mutate} = testClientPrivate
      const input: CreatePeerInput = {
        name: 'Create Peer Test',
        slug: 'test-peer',
        hostURL: 'https://host-url.ch/',
        token: 'testTokenABC123'
      }
      const res = await mutate({
        mutation: CreatePeer,
        variables: {
          input: input
        }
      })
      expect(res).toMatchSnapshot({
        data: {
          createPeer: {
            id: expect.any(String)
          }
        }
      })
      ids.unshift(res.data?.createPeer?.id)
    })

    test('can be read in list', async () => {
      const {query} = testClientPrivate
      const res = await query({
        query: PeerList
      })
      expect(res).toMatchSnapshot({
        data: {
          peers: Array.from({length: ids.length}, () => ({
            id: expect.any(String)
          }))
        }
      })
      expect(res.data?.peers?.length).toBe(ids.length)
    })

    test('can be read by id', async () => {
      const {query} = testClientPrivate
      const res = await query({
        query: Peer,
        variables: {
          id: ids[0]
        }
      })
      expect(res).toMatchSnapshot({
        data: {
          peer: {
            id: expect.any(String)
          }
        }
      })
    })

    test('can be updated', async () => {
      const {mutate} = testClientPrivate
      const res = await mutate({
        mutation: UpdatePeer,
        variables: {
          input: {
            name: 'Updated Peer',
            slug: 'update-peer',
            hostURL: 'https://new-host-url.ch/',
            token: 'newTestTokenABC123'
          },
          id: ids[0]
        }
      })
      expect(res).toMatchSnapshot({
        data: {
          updatePeer: {
            id: expect.any(String)
          }
        }
      })
    })

    test('can be deleted', async () => {
      const {mutate} = testClientPrivate
      const res = await mutate({
        mutation: DeletePeer,
        variables: {
          id: ids[0]
        }
      })
      expect(res).toMatchSnapshot({
        data: {
          deletePeer: expect.any(String)
        }
      })
      ids.shift()
    })

    test('can read peer profile', async () => {
      const {query} = testClientPrivate
      const res = await query({
        query: PeerProfile
      })
      expect(res).toMatchSnapshot()
    })

    test('can update peer profile', async () => {
      const {mutate} = testClientPrivate
      const input: PeerProfileInput = {
        name: 'test peer profile',
        logoID: 'logoID123',
        themeColor: '#4287f5',
        callToActionText: [{text: 'rich text call to action'}],
        callToActionURL: 'calltoactionurl.ch/'
      }
      const res = await mutate({
        mutation: UpdatePeerProfile,
        variables: {
          input: input
        }
      })
      expect(res).toMatchSnapshot()
    })
  })
})

afterAll(async () => {
  if (dbAdapter) {
    await dbAdapter.db.dropDatabase()
    await dbAdapter.client.close()
  }
})
