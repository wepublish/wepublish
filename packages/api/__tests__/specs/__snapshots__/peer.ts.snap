// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Peers can be created/edited/deleted: can be created 1`] = `
Object {
  "data": Object {
    "createPeer": Object {
      "hostURL": "https://host-url.ch/",
      "id": Any<String>,
      "name": "Create Peer Test",
      "slug": "test-peer",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Peers can be created/edited/deleted: can be deleted 1`] = `
Object {
  "data": Object {
    "deletePeer": Any<String>,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Peers can be created/edited/deleted: can be read by id 1`] = `
Object {
  "data": Object {
    "peer": Object {
      "hostURL": "https://host-url.ch/",
      "id": Any<String>,
      "name": "Peer Test 3",
      "slug": "test-peer3",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Peers can be created/edited/deleted: can be read in list 1`] = `
Object {
  "data": Object {
    "peers": Array [
      Object {
        "hostURL": "https://host-url.ch/",
        "id": Any<String>,
        "name": "Peer Test 2",
        "slug": "test-peer2",
      },
      Object {
        "hostURL": "https://host-url.ch/",
        "id": Any<String>,
        "name": "Create Peer Test",
        "slug": "test-peer",
      },
      Object {
        "hostURL": "https://host-url.ch/",
        "id": Any<String>,
        "name": "Peer Test 0",
        "slug": "test-peer0",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Peers can be created/edited/deleted: can be updated 1`] = `
Object {
  "data": Object {
    "updatePeer": Object {
      "hostURL": "https://new-host-url.ch/",
      "id": Any<String>,
      "name": "Updated Peer",
      "slug": "update-peer",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Peers can be created/edited/deleted: can read peer profile 1`] = `
Object {
  "data": Object {
    "peerProfile": Object {
      "callToActionImage": null,
      "callToActionImageURL": "",
      "callToActionText": Array [],
      "callToActionURL": "",
      "hostURL": "https://fakeURL",
      "logo": null,
      "name": "",
      "themeColor": "#000000",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Peers can be created/edited/deleted: can update peer profile 1`] = `
Object {
  "data": Object {
    "updatePeerProfile": Object {
      "callToActionImage": null,
      "callToActionImageURL": null,
      "callToActionText": Array [
        Object {
          "text": "rich text call to action",
        },
      ],
      "callToActionURL": "calltoactionurl.ch/",
      "hostURL": "https://fakeURL",
      "logo": null,
      "name": "test peer profile",
      "themeColor": "#4287f5",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
